src/guid.rs

```diff
@@ -0,0 +1,123 @@
+use std::{fmt, str, ops};
+
+// TODO: We could store a 13th b'\0' byte for a C-compatible null terminator. Is
+// this worth doing?
+#[derive(Clone, Copy, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct Guid([u8; 12]);
+
+// These are here so that we can create these as const/static (without needing lazy_static)
+pub(crate) const ROOT_GUID: Guid = Guid(*b"root________");
+
+pub(crate) const USER_CONTENT_ROOTS: [Guid; 4] = [
+    Guid(*b"toolbar_____"),
+    Guid(*b"menu________"),
+    Guid(*b"unfiled_____"),
+    Guid(*b"mobile______"),
+];
+
+impl Guid {
+    #[inline]
+    pub fn try_new(s: &str) -> Option<Guid> {
+        if !Guid::is_valid(s) {
+            None
+        } else {
+            assert!(s.is_char_boundary(12));
```

Comment:
Why not `s.len() == 12`?

Reply:
This is the absolute minimum requirement we would need to maintain to make using `str::from_utf8_unchecked` in `as_str` safe.